src/yaml.rs

```diff
@@ -282,13 +325,82 @@ impl<T: std::io::Read> YamlDecoder<T> {
         let mut buffer = Vec::new();
         self.source.read_to_end(&mut buffer)?;
 
-        // Decodes the input buffer using either UTF-8, UTF-16LE or UTF-16BE depending on the BOM codepoint.
-        // If the buffer doesn't start with a BOM codepoint, it will use a fallback encoding obtained by
-        // detect_utf16_endianness.
-        let (res, _) =
-            encoding::types::decode(&buffer, self.trap, detect_utf16_endianness(&buffer));
-        let s = res.map_err(LoadError::Decode)?;
-        YamlLoader::load_from_str(&s).map_err(LoadError::Scan)
+        // Check if the `encoding` library can detect encoding from the BOM, otherwise use
+        // `detect_utf16_endianness`.
+        let (encoding, _) =
+            Encoding::for_bom(&buffer).unwrap_or_else(|| (detect_utf16_endianness(&buffer), 2));
+        let mut decoder = encoding.new_decoder();
+        let mut output = String::new();
+
+        // Decode the input buffer.
+        decode_loop(&buffer, &mut output, &mut decoder, self.trap)?;
+
+        YamlLoader::load_from_str(&output).map_err(LoadError::Scan)
+    }
+}
+
+/// Perform a loop of [`Decoder::decode_to_string`], reallocating `output` if needed.
+fn decode_loop(
+    input: &[u8],
+    output: &mut String,
+    decoder: &mut Decoder,
+    trap: YAMLDecodingTrap,
+) -> Result<(), LoadError> {
+    output.reserve(input.len());
+    let mut total_bytes_read = 0;
+
+    loop {
+        match decoder.decode_to_string_without_replacement(&input[total_bytes_read..], output, true)
+        {
+            // If the input is empty, we processed the whole input.
+            (DecoderResult::InputEmpty, _) => break Ok(()),
+            // If the output is full, we must reallocate.
+            (DecoderResult::OutputFull, bytes_read) => {
+                total_bytes_read += bytes_read;
+                output.reserve(input.len() / 10);
```

Comment:
I'd lean towards removing the ` / 10` here since we're just reserving. It should lead to less reallocs overall. Another common strategy for really large buffers is to double the reserve() size each time we hit OutputFull.

Reply:
The output is already reserved to the size of the input. Would it be common to have instances of inputs which, when converted to utf8, are more than 1.1x their size? 

This definitely needs a `max` though. If the input size is less than 10 bytes that's a reserve of 0.

Reply:
Ah, that's true. Most of the time we'd probably expect something encoded in utf32 to become smaller when converted to utf8, so the `/ 10` makes sense, especially with a short comment to explain the magic number. `/ 4` would give a safe buffer for some pathological cases, but in light of this is probably overkill. I wonder whether this code path is ever hit in the wild.

Reply:
Added a comment, thanks!

> I wonder whether this code path is ever hit in the wild.

I know nearly nothing about encodings, but I'm curious :thinking: